faux-mgs/src/host_phase2.rs

```diff
@@ -0,0 +1,188 @@
+                    return None;
+                }
+            };
+
+            if file_type.is_symlink() {
+                // Attempt to check whether the linked-to file is a file or
+                // directory.
+                let meta = match fs::metadata(&path).await {
+                    Ok(meta) => meta,
+                    Err(err) => {
+                        warn!(
+                            log, "failed to read metadata of {}", path.display();
+                            "err" => %err,
+                        );
+                        return None;
+                    }
+                };
+                meta.file_type().is_file().then_some(path)
+            } else {
+                file_type.is_file().then_some(path)
+            }
+        }).filter_map(convert::identity);
+
+        // For all these path bufs, open them and compute hashes.
+        let file_hashes = dir_files
+            .then(|path| async move {
+                let mut file = match File::open(&path).await {
+                    Ok(file) => file,
+                    Err(err) => {
+                        warn!(
+                            log, "failed to open {}", path.display();
+                            "err" => %err,
+                        );
+                        return None;
+                    }
+                };
+
+                // TODO: When the host requests a phase 2 image by hash, it's
+                // the hash of the file _after_ its 4 KiB header (that header
+                // includes the hash itself). For now in faux-mgs, we'll just
+                // skip over the first 4 KiB when computing the hash of files
+                // we're going to serve. In real MGS, we probably want to do
+                // something smarter here: maybe compute the hash (skipping 4
+                // KiB) and then check against the hash in the header itself?
+                match file.seek(SeekFrom::Start(4096)).await {
+                    Ok(_) => (),
+                    Err(err) => {
+                        warn!(
+                            log, "failed to seek to offset 4096 in {}",
+                            path.display();
+                            "err" => %err,
+                        );
+                        return None;
+                    }
+                }
+
+                let mut hasher = Sha256::new();
+                let mut buf = [0; 4096];
+                let digest = loop {
+                    match file.read(&mut buf).await {
+                        Ok(0) => break hasher.finalize(),
+                        Ok(n) => hasher.update(&buf[..n]),
+                        Err(err) => {
+                            warn!(
+                                log, "error reading {}", path.display();
+                                "err" => %err,
+                            );
+                            return None;
+                        }
+                    }
+                };
```

Comment:
This was _slightly_ more involved because `file` here is async and `hasher` is not, but it let me go poking around in how to wrap a sync writer in an async compatibility layer, so learned something new. Changed in 91b760d543bfb48167d0d365ff88dbcad2b07809

Reply:
Interesting!